File: src/main/secrets.ts
import { safeStorage } from 'electron'
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'

/**
 * Stores secrets (passwords/passphrases) encrypted with the OS keystore via
 * Electron's safeStorage (DPAPI on Windows). Only ciphertext is written to
 * disk; plaintext is never persisted. Keyed by an opaque id (e.g.
 * `user@host:port`).
 */
export class SecretStore {
  private data: Record<string, string> = {} // id -> base64 ciphertext

  constructor(private readonly filePath: string) {
    this.load()
  }

  /** Whether OS-backed encryption is usable. If false, we refuse to persist. */
  get available(): boolean {
    return safeStorage.isEncryptionAvailable()
  }

  private load(): void {
    try {
      if (existsSync(this.filePath)) {
        this.data = JSON.parse(readFileSync(this.filePath, 'utf8')) as Record<string, string>
      }
    } catch {
      this.data = {}
    }
  }

  private save(): void {
    mkdirSync(dirname(this.filePath), { recursive: true })
    writeFileSync(this.filePath, JSON.stringify(this.data, null, 2), { mode: 0o600 })
  }

  has(id: string): boolean {
    return id in this.data
  }

  /** Encrypt and persist a secret. Returns false if encryption is unavailable. */
  set(id: string, plaintext: string): boolean {
    if (!this.available) return false
    this.data[id] = safeStorage.encryptString(plaintext).toString('base64')
    this.save()
    return true
  }

  /** Decrypt a stored secret, or null if absent/unavailable/corrupt. */
  get(id: string): string | null {
    if (!this.available) return null
    const ciphertext = this.data[id]
    if (!ciphertext) return null
    try {
      return safeStorage.decryptString(Buffer.from(ciphertext, 'base64'))
    } catch {
      return null
    }
  }

  delete(id: string): void {
    if (id in this.data) {
      delete this.data[id]
      this.save()
    }
  }
}

File: src-tauri/src/secrets.rs
//! Stores secrets (passwords/passphrases) in the OS keystore via the keyring
//! crate (Windows Credential Manager on Windows). Plaintext never crosses to
//! the renderer — only presence checks and forget; decrypted values are
//! injected into connect requests backend-side.
//!
//! Ids use the same scheme as the Electron build: `pw:user@host:port` for
//! passwords, `pp:<keyPath>` for key passphrases. (Old DPAPI-encrypted
//! secrets.json files are intentionally not migrated.)

const SERVICE: &str = "simple-ssh";

/// Stable id for a stored password: pw:user@host:port.
pub fn password_id(host: &str, port: u16, username: &str) -> String {
    format!("pw:{username}@{host}:{port}")
}

/// Stable id for a stored key passphrase, keyed by key path.
pub fn passphrase_id(key_path: &str) -> String {
    format!("pp:{key_path}")
}

fn entry(id: &str) -> Result<keyring::Entry, keyring::Error> {
    keyring::Entry::new(SERVICE, id)
}

pub struct SecretStore;

impl SecretStore {
    pub fn has(&self, id: &str) -> bool {
        entry(id).and_then(|e| e.get_password()).is_ok()
    }

    /// Persist a secret. Returns false if the OS keystore rejected it.
    pub fn set(&self, id: &str, plaintext: &str) -> bool {
        entry(id).and_then(|e| e.set_password(plaintext)).is_ok()
    }

    /// Fetch a stored secret, or None if absent/unavailable.
    pub fn get(&self, id: &str) -> Option<String> {
        entry(id).and_then(|e| e.get_password()).ok()
    }

    pub fn delete(&self, id: &str) {
        let _ = entry(id).and_then(|e| e.delete_credential());
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn id_schemes_match_electron() {
        assert_eq!(password_id("example.com", 22, "alice"), "pw:alice@example.com:22");
        assert_eq!(passphrase_id("C:\\keys\\id_ed25519"), "pp:C:\\keys\\id_ed25519");
    }
}
